bin/docker.sh

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+# Author: Valentin Kuznetsov <vkuznet@gmail.com>
+# helper script to handle docker actions
+
+if [ $# -ne 4 ]; then
+     echo "Usage: docker.sh <action> <service> <tag> <registry> "
+     echo "Supported actions: build, build-stable, push"
+     exit 1;
+fi
+
+# helper function to match the pattern X.Y.Z where X, Y, and Z are numbers
+function match_tag {
+  local tag=$1
+  if [[ $tag =~ ^([0-9]+)\.([0-9]+)\.([0-9]+)$ ]]; then
+    return 0
+  elif [[ $tag =~ ^([0-9]+)\.([0-9]+)\.([0-9]+)\.([0-9]+)$ ]]; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+# make input variable assignments
+action=$1
+service=$2
+tag=$3
+registry=$4
+rurl=$registry/cmsweb/$service
+
+# check if action is supported
+case "$action" in
+  "build"|"build-stable"|"push"|"remove"|"delete")
+    ;;
+  *)
+    echo "action: '$action' is not supported"
+    exit 1
+    ;;
+esac
+
+# determine suffix of image tag
+suffix=""
+if match_tag "$tag"; then
+    suffix="-stable"
+fi
+
+# check build/push actions and build the image(s) if it will be required
+if [ "$action" == "build" ]; then
+    echo "check if image with service=$service and tag=$tag exist ..."
+    image_exist=`docker images | grep $tag | grep $service`
+    if [ -z "$image_exist" ]; then
+        echo "action: docker build --build-arg TAG=${tag} --tag ${rurl}:${tag}"
+        docker build --build-arg TAG=${tag} --tag ${rurl}:${tag} .
+    fi
+    docker images | grep $tag | grep $service
+fi
+if [ "$action" == "build-stable" ] || ( [ "$action" == "build" ] && [ -n "$suffix" ] ); then
+    if [ -z "$suffix" ]; then
+        echo "Building stable image for tag=$tag is not appropriate as tag is not matched X.Y.Z or X.Y.Z.P pattern"
+        exit 0
+    fi
+    echo "check if image with service=$service and tag=${tag}${suffix} exist ..."
+    image_exist=`docker images | grep ${tag}${suffix} | grep $service`
+    if [ -z "$image_exist" ]; then
+        echo "action: docker build --build-arg TAG=${tag} --tag ${rurl}:${tag}${suffix}"
+        docker build --build-arg TAG=${tag} --tag ${rurl}:${tag}${suffix} .
+    fi
+    docker images | grep ${tag}${suffix} | grep $service
+fi
+if [ "$action" == "push" ] || ( [ "$action" == "push" ] && [ -n "$suffix" ] ); then
+    image_exist=`docker images | grep $tag | grep $service`
```

Comment:
This line can potentially list $tag and $tag$suffix. Meaning that the if statement below might be ill evaluated (and the `echo` message needs to be corrected as well (without $suffix).

Reply:
I see.
The code seems to be functional, but that 3 level nested `if` can probably be refactored as well.
Here is the logic I would suggest implementing:
1. check if image exist (filter out the stable one)
2. if it does not, exit with an ERROR, else docker push tag
3. if suffix is not empty, check if stable image exist
4. if it does not, exit with an ERROR, else docker push stable tag

I would say it's just a matter of moving your lines of code here and there and you can have this with a single level of if statements. Perhaps a code like (incomplete):
```
    image_exist=`docker images | grep $tag | grep $service | grep -v ${tag}${suffix}`
    image_exist_stable=`docker images | grep ${tag}${suffix} | grep $service`
    if [ -z "$image_exist" ]; then
        echo "Images ${rurl}:${tag} not found"
        exit 1
    fi
    if [ -n "$suffix" ] && [ -z "$image_exist_stable" ]; then  # FIXME please check the syntax
        echo "Images ${rurl}:${tag}${suffix} not found"
        exit 2
    fi
    echo "action: docker push ${rurl}:${tag}"
    docker push ${rurl}:${tag}
    echo "action: docker push ${rurl}:${tag}${suffix}"
    docker push ${rurl}:${tag}${suffix}
```

Reply:
Sounds reasonable to me. This syntax doesn't push the `-stable` image if the `a.b.c.` image doesn't exist, but given the way this script is used, I think that this behavior is fine, for the time being.